More documentation (#377)
OCamlPro/superbol-studio-oss: README.md

> (Temporary limitation)
>
> Suggestions of user-defined words may not comprise symbols that are

> > defined in some sections of the data division.  They do not include

Reviewer: I really do wonder what "some sections" means... also it would be reasonable to say here something like "user-defined words from configured copybooks [link] are included" (if this is the case, what I guess).

> ![Go to Definition](./assets/superbol-goto-definition.gif)

Reviewer: One of the navigation features of vscode is "go to symbol in workspace" - is this supported and if not hard to add (#365)?

Author: No this is not supported yet. This apparently requires some ability to precisely scan a whole workspace, which we anticipate to be a rather costly operation, and therefore needs to be designed with care on the LSP server's side to avoid too many noticeable impact on performance.

> [!NOTE]
> (Temporary limitation)
>
> At the moment, definitions that belong to *some* sections of the
> data division, and some references to data items, are ignored by the

> > extension.  They will be covered by the first stable release.

Reviewer 2: ```suggestion
> extension. They will be covered by the first stable release.
```

> > extension.  They will be covered by the first stable release.

Reviewer 3: ```suggestion
> extension. They will be covered by the first stable release.
```

> [!NOTE]
> (Temporary limitation)
>
> In addition to limitations mentioned in [Go to
> Definition](#go-to-definition), references that occur in
> `EXEC`/`END-EXEC` blocks are also not taken into account yet.

### Reference Information

The extension shows inline reference information above definitons of
data items and elements of the procedure division.  
The same limitations as for [Go to Definition](#go-to-definition)

> apply.

Reviewer: please add a sscreenshot showing the feature here as well; I _guess_ that possibly means #354 can be closed

Author: Thanks indeed the screenshot is missing here.
#354 would probably be easy to implement (including the reference count in the generated information shown on hover), but it's not done yet.

![Hover over replacement](./assets/superbol-hover-replacement.gif)

## Editing

### Rename Data Items, Sections, and Paragraphs

You can rename any data item by pressing <kbd>F2</kbd> while your
cursor is positioned on one of its references.  The extension will
warn you if a reference to the renamed item appears in a copybook (in
which case the renaming of every reference is not performed).

![Rename Symbol](./assets/superbol-rename-symbol.gif)

![Rename Symbol in Copybook](./assets/superbol-rename-symbol-in-copybook.gif)

Sections and paragraphs of the procedure division can also be renamed
in the same way.  
The same limitations as for [Go to Definition](#go-to-definition)
apply to this feature.

## Exploring the Control-flow

Navigating a graphical representation of a COBOL program's
control-flow proves invaluable when in comes to decypher its overall

> logic.  To do this, open the command palette (or type

Reviewer 2: ```suggestion
logic. To do this, open the command palette (or type
```

<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd>), and select `SuperBOL:
Show control-flow` (you can also right click and select `Show

> control-flow` in the menu).  You are then presented with a list of

Reviewer 2: ```suggestion
control-flow` in the menu). You are then presented with a list of
```

portions of program to consider (either the entire program, or
individiual sections): select one element to see the corresponding
CFG.

![CFG Explorer](./assets/superbol-cfg-explorer.gif)

Various settings are provided to tune the rendering of CFGs.

![CFG Explorer Collapse Fallthrough](./assets/superbol-cfg-explorer-collapse-fallthrough.gif)

### CFG as an arc diagram

> A rendering of CFGs as arc diagrams is also available.  In this

Reviewer 2: ```suggestion
A rendering of CFGs as arc diagrams is also available. In this
```

> representation, sections and paragraph named are layed out vertically,

Reviewer 3: ```suggestion
representation, sections and paragraphs named are laid out vertically,
```

in the same way.  
The same limitations as for [Go to References](#go-to-references)
apply to this feature.

## Exploring the Control-flow

Navigating a graphical representation of a COBOL program's
control-flow proves invaluable when it comes to decipher its overall
logic.  To do this, open the command palette (or type
<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>P</kbd>), and select `SuperBOL:
Show control-flow` (you can also right click and select `Show
control-flow` in the menu).  You are then presented with a list of
portions of program to consider (either the entire program, or
individiual sections): select one element to see the corresponding
CFG.

![CFG Explorer](./assets/superbol-cfg-explorer.gif)

Various settings are provided to tune the rendering of CFGs.

![CFG Explorer Collapse Fallthrough](./assets/superbol-cfg-explorer-collapse-fallthrough.gif)

### CFG as an arc diagram

A rendering of CFGs as arc diagrams is also available.  In this
representation, named sections and paragraphs are laid out vertically,
and arcs show the direction of control-flow between them.

![CFG Explorer as Arc Diagram](./assets/superbol-cfg-explorer-arc.gif)

## Debugging

> In order to debug a COBOL program, you first need to run a *build

Reviewer: This section misses things like attaching, setting program/procedure breakpoints, watchpoints, ...

If you change the extension to keep the debugging in its own separate extension (with a dependency in this extension to the debug extension but not vice-versa if the other extension does not use the LSP), then you'd only need a teaser here and can reference to that's extension's README, which should already include parts of this.

Also I'd highly suggest to not only hint to the debug build task, because that is only useful for small/testing programs [
but there it is _very_ useful, especially for beginners]. With bigger environments you'll have either external build flows or (most preferably) done it by `make`.

For this scenario it is important to document which options need to be passed to cobc (I _think_ for GC 3.2+ the `-g` option is enough).

As the coverage handling is also part of that extension, the same applies there - but in any case I'd suggest to have at least a screenshot of that here.

Author: I've added basic details about the additional flags passed to `cobc`.

Regarding dependencies, at the moment the debug extension is tuned with a `preLaunchTask` that directs to a build task of the main extension; this is mainly to help newcomers (they can just press <kbd>F5</kbd> to automatically compile and run programs). I'd like to keep the extensions separate, but for now we found no convincing way to do that and keep the aforementioned behavior (it's also still debatable whether this fully automated behavior is really worth the hassle — see also the other debug-related issues).

At the moment I'm trying to do a quick release (long overdue); so aspects mentioned in this comment will be addressed later.

Note we don't rely on the debug extension to show coverage information (there's no reason to have those two features so tied together I think): a dedicated gcov-viewer extension that we found simpler/more maintainable than the coverage-related code of the debug extension is used instead.

Reviewer: > a dedicated gcov-viewer extension that we found simpler/more maintainable than the coverage-related code of the debug extension is used instead.

That's good - can you keep that in a separate (depending = automatically installed [as long as you're connected to the internet]) extension or does that needs to "included" as well?

Please create an issue to drop the coverage stuff from the debugging extension (because it is not needed any more and is "too much" there as well).
